### DISCUSSION TITLE
Split e2e suite for app into separate, smaller suites for features within app

### DISCUSSION THREAD

**Discussion by wrslatz** [nrwl/nx#7668] [Feature Requests]
### Overview
We currently use Cypress to test our application. I'm curious if there is an opportunity to better support more granular e2e suites to leverage affected analysis for more performant e2e testing. We currently have a manual setup for this, but it seems like an opportunity to bring more performant e2e suites with better affected analysis as a supported feature by Nx.

### Our current setup
1. Separate e2e suite for each feature (often a single page or group of related pages) we want to e2e test, with the e2e suite having the feature library in it's `implicitDependencies`
2. Using `start-server-and-test` to run the app and then run all `e2e` targets affected by a particular change.
3. We have a main e2e suite testing common functionality across the entire app

### Our current pain points
1. No generator for Cypress e2e suites makes spinning up new ones a pain (we either generate a new app and then delete the app leaving it's e2e or copy an existing e2e and replace)
2. `start-server-and-test` has issues chaining arguments, making it cumbersome to use with `nx` CLI. We'd prefer if Nx supported reusing and then tearing down a single `devServerTarget` for a group of e2e suites (related issue https://github.com/nrwl/nx/issues/6835)

**Comment by brandonroberts** [comment]
@wrslatz we are looking at documenting this better, but you can create an E2E project from an existing app. Quoted from @Coly010 in our Community Slack

```
if the @nrwl/cypress package doesn't exist, first run:
npm install --save-dev @nrwl/cypress
once that is done run:
nx g @nrwl/cypress:cypress-project appName-e2e --project=appName
replacing appName with the app's name as defined in your angular.json file
```

**Reply by Coly010** [reply]
Following on from this, theoretically (I haven't tried it), once you have multiple e2e projects that all require one app, you could set up a custom executor on the app's project that will serve the app once and then run the tests from each of e2e projects against it.

**Reply by wrslatz** [reply]
Thanks for the super fast reply! 🚀 

> you can create an E2E project from an existing app...

I'll give that method a try. Would it makes sense to expose that generator in the Nx Console? Maybe that would pair well with documentation improvements.

**Reply by wrslatz** [reply]
> Following on from this, theoretically (I haven't tried it), once you have multiple e2e projects that all require one app, you could set up a custom executor on the app's project that will serve the app once and then run the tests from each of e2e projects against it.

https://github.com/nrwl/nx/issues/3748 (linked in https://github.com/nrwl/nx/issues/6835 as related) calls out the main issue with the custom executor approach: the serve target(s) that all of the e2e depend on don't get stopped when the e2e targets finish.

That was my main motivation for raising https://github.com/nrwl/nx/issues/6835; I think if Nx supported starting and cleaning up dependent serve targets, it would make it much easier to setup these more performant affected e2e scenarios and help close out several of these issues.

**Reply by wrslatz** [reply]
Reading deeper on the thread in https://github.com/nrwl/nx/issues/3748, it seems like serving a web target _may_ handle closing the server after tests run. I'll give that a try soon 🤞

**Reply by wrslatz** [reply]
I tried a custom `nx:run-commands` executor to serve the app once. Using `devServerTarget`, the custom serve executor is invoked on execution of each Cypress suite, so that doesn't work.

Using `dependsOn`, I couldn't get the custom `nx:run-commands` executor to be run correctly at all (I probably had something misconfigured or misunderstood, or this isn't supported yet). https://github.com/nrwl/nx/issues/5570 seems somewhat related to this path.

Looks like for now we have to use `start-server-and-test`, which doesn't support arg chaining so it leads to a poorer DX.

**Comment by mirobo** [comment]
I wonder why there aren't more people writing about this :D. I have the same issue: Lots of tests with parallel execution but still the execution time will only grow if we do not split tests into smaller chunks. So far I've seen some quite ok examples in various threads here. 

But imagine if a single Cypress test is added in a spec file in a certain library or a new spec file is added. At the moment I'm not sure if it would even execute the e2e tests, if a new one was added.

I would expect it to execute all specs that belong to this "app" (because by default e2e tests are an "app"). This leads to the next problem: IMO it should only execute the changed/added specs within this app and not everything. I haven't seen any code that would narrow down the tests to execute via the Cypress "specPattern" param. I think I would need to do this somehow with a custom executor?

At this point I ask myself why I would use "nx affected" to narrow tests down via the project graph (on a per library basis) and additionally add single tests that were changed but were not "seen" by the project graph.

Maybe it's better to treat e2e tests as a library too?

I'll probably setup a new nx test project. The calculation on what is affected is quite a blackbox (without spending a LOT of time studying the code) and local experiments with "npx nx print-affected --type=app --select=projects --uncommitted=true" were not really helpful.. and didn't work as expected (sometimes it wouldn't show anything and if I removed "--select=projects" it would dump out the whole project graph instead of what I actually want to see: The affected libs and apps.

I was also looking into tools like https://github.com/pahen/madge , https://github.com/sverweij/dependency-cruiser and https://github.com/vladimirnani/typescript-class-usage-graph ..but I guess this will get extremly complicated and probably also slow. So I guess after all the NX project setup is the best possible solution...
